File: c++/klotski_solver.cpp
// SPDX-License-Identifier: MIT
// Copyright © 2023  Bilal Djelassi

#include "puzzle_types.hpp"
#include "svg_renderer.hpp"
#include "xml_writer.hpp"
#include <algorithm>
#include <format>
#include <fstream>
#include <functional>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <vector>


using KlotskiGrid = Grid<4, 5>;

struct SearchEdge
{
    size_t parentIndex;
    Move move;
};

using KlotskiSearchTree =
    SearchTree<KlotskiGrid, std::optional<SearchEdge>, std::string>;

struct KlotskiSolution
{
    KlotskiGrid grid;
    std::vector<Move> path;
};

KlotskiSolution solvePuzzle(
    KlotskiGrid const& initialGrid,
    std::function<bool (KlotskiGrid const&)> successCondition,
    KlotskiGrid::KeySymmetry symmetry)
{
    auto const validated = initialGrid.validate();
    if (!validated)
        throw std::runtime_error("initial grid is invalid");

    if (successCondition(initialGrid))
        return { initialGrid, {} };

    KlotskiSearchTree searchTree;
    searchTree.append(initialGrid, std::nullopt, validated->key(symmetry));

    while (true) {
        searchTree.incrementDepth();
        auto const indexRange = searchTree.currentDepth();

        if (indexRange.isEmpty())
            throw std::runtime_error("reached end of tree, no more solutions to explore");

        // loop over last reached grids ...
        for (size_t const parentIndex : indexRange)

        // ... for each piece ...
        for (size_t const pieceIndex : IndexRange{0, initialGrid.pieces.size()})

        // ... and by trying each step as a move
        for (auto const& step : Step::all()) {
            auto const move = Move{pieceIndex, step};
            KlotskiGrid grid = searchTree.nodeAt(parentIndex);
            grid.apply(move);

            auto const validated = grid.validate();
            if (!validated)
                continue;

            auto const key = validated->key(symmetry);
            if (!searchTree.append(grid, SearchEdge{parentIndex, move}, key))
                continue;

            if (successCondition(grid)) {
                KlotskiSolution solution = {};
                solution.grid = grid;

                for (auto edge = searchTree.edgeAt(searchTree.lastIndex());
                          edge != std::nullopt;
                          edge = searchTree.edgeAt(edge->parentIndex))
                    solution.path.push_back(edge->move);

                std::reverse(solution.path.begin(), solution.path.end());
                return solution;
            }
        }
    }
}

struct KlotskiSVGRenderer : public SVGRenderer<KlotskiGrid>
{
    KlotskiSVGRenderer() {
        properties = colorfulTheme();
        pieceDefs = {
            { 'A', 2 * unitSize + cellGap, 2 * unitSize + cellGap },
            { 'B', unitSize, 2 * unitSize + cellGap },
            { 'C', 2 * unitSize + cellGap, unitSize },
            { 'D', unitSize, unitSize },
        };
        stepDefs = {
            { Step::up().vector,    "stepUp",    "M 0 -6 L -6 4 L 6 4 Z" },
            { Step::down().vector,  "stepDown",  "M 0 6 L -6 -4 L 6 -4 Z" },
            { Step::left().vector,  "stepLeft",  "M -6 0 L 4 -6 L 4 6 Z" },
            { Step::right().vector, "stepRight", "M 6 0 L -4 -6 L -4 6 Z" },
        };
    }

private:
    struct PieceDef {
        char symbol;
        float sizeX;
        float sizeY;
    };
    struct StepDef {
        Vect2 vector;
        std::string id;
        std::string path;
    };

    std::vector<PieceDef> pieceDefs;
    std::vector<StepDef> stepDefs;

protected:
    PieceDef const& getPieceDef(char symbol) const {
        auto iter = std::find_if(pieceDefs.cbegin(), pieceDefs.cend(),
        [&](PieceDef const& pieceDef) {
            return pieceDef.symbol == symbol;
        });
        if (iter == pieceDefs.cend())
            throw std::runtime_error("cannot access requested piece definition");
        return *iter;
    }

    StepDef const& getStepDef(Vect2 const& vector) const {
        auto iter = std::find_if(stepDefs.cbegin(), stepDefs.cend(),
        [&](StepDef const& stepDef) {
            return stepDef.vector == vector;
        });
        if (iter == stepDefs.cend())
            throw std::runtime_error("cannot access requested step definition");
        return *iter;
    }

    void preRender(XmlWriter& svg, KlotskiGrid const& grid) const override {
        using std::format;
        svg.elem("defs");

        // title style
        svg.elem("style")
           .attr("type", "text/css")
           .text(format(
              "text.title {{"
              "text-anchor: middle;"
              "fill: {};"
              "font-family: {};"
              "font-size: {};"
              "}}",
               property("textColor"),
               property("fontFamily"),
               property("fontSize")))
           .term();

        // prerender grid
        svg.elem("rect")
           .attr("id", "grid")
           .attr("x", "0")
           .attr("y", "0")
           .attr("width", format("{}", getGridSizeX()))
           .attr("height", format("{}", getGridSizeY(false)))
           .attr("rx", format("{}", borderRadius))
           .attr("fill", property("fillColorGrid"))
           .attr("stroke", property("strokeColorGrid"))
           .attr("stroke-width", property("strokeWidthGrid"))
           .term();

        // if grid has obstacles, include prerendered obstacle
        if (!grid.obstacles.empty()) {
            svg.elem("pattern")
               .attr("id", "hachures")
               .attr("width", "4")
               .attr("height", "4")
               .attr("patternUnits", "userSpaceOnUse")
               .attr("patternTransform", "rotate(-45)")
                  .elem("line")
                  .attr("x1", "0")
                  .attr("y1", "1")
                  .attr("x2", "4")
                  .attr("y2", "1")
                  .attr("stroke", property("strokeColorGrid"))
                  .attr("stroke-width", "2")
                  .term()
               .term()
               .elem("rect")
               .attr("id", "obstacle")
               .attr("width", format("{}", unitSize))
               .attr("height", format("{}", unitSize))
               .attr("rx", format("{}", borderRadius))
               .attr("fill", "use(#hachures)")
               .term();
        }

        // prerender pieces and steps
        for (auto const& pieceDef : pieceDefs) {
            svg.elem("rect")
               .attr("id", format("piece{}", pieceDef.symbol))
               .attr("x", "0")
               .attr("y", "0")
               .attr("width", format("{}", pieceDef.sizeX))
               .attr("height", format("{}", pieceDef.sizeY))
               .attr("rx", format("{}", borderRadius))
               .attr("fill", property(format("fillColorLoPiece{}", pieceDef.symbol)))
               .attr("stroke", property(format("strokeColorLoPiece{}", pieceDef.symbol)))
               .attr("stroke-width", property("strokeWidthPiece"))
               .term()
               .elem("rect")
               .attr("id", format("piece{}_", pieceDef.symbol))
               .attr("x", "0")
               .attr("y", "0")
               .attr("width", format("{}", pieceDef.sizeX))
               .attr("height", format("{}", pieceDef.sizeY))
               .attr("rx", format("{}", borderRadius))
               .attr("fill", property(format("fillColorHiPiece{}", pieceDef.symbol)))
               .attr("stroke", property(format("strokeColorHiPiece{}", pieceDef.symbol)))
               .attr("stroke-width", property("strokeWidthPiece"))
               .term();
        }

        float const stepScale = unitSize / 20.0f;
        for (auto const& stepDef : stepDefs) {
            svg.elem("path")
               .attr("id", stepDef.id)
               .attr("d", stepDef.path)
               .attr("fill", property("arrowColor"));
            if (stepScale != 1.0f)
                svg.attr("transform", format("scale({})", stepScale));
            svg.term();
        }
        svg.term();
    }

    void render(XmlWriter& svg, KlotskiGrid const&,
        std::optional<std::string> title) const override {

        using std::format;

        svg.elem("use")
           .attr("xlink:href", "#grid")
           .term();

        if (title) {
            svg.elem("text")
               .attr("x", format("{}", getGridSizeX() * 0.5f))
               .attr("y", format("{}", titleHeight * -0.5f))
               .attr("class", "title")
               .text(title.value())
               .term();
        }
    }

    void render(XmlWriter& svg, Piece const& piece,
        std::optional<Step> step) const override {

        using std::format;

        std::string const pieceId =
            format("piece{}{}", piece.tag.symbol, step ? "_" : "");

        float const piecePosX = piecePositionToXCoord(piece.position.x);
        float const piecePosY = piecePositionToYCoord(piece.position.y);

        svg.elem("use")
           .attr("xlink:href", format("#{}", pieceId))
           .attr("transform", format("translate({} {})", piecePosX, piecePosY))
           .term();

        if (step) {
            StepDef const& stepDef = getStepDef(step->vector);
            PieceDef const& pieceDef = getPieceDef(piece.tag.symbol);

            float const stepPosX = piecePosX + pieceDef.sizeX * 0.5f;
            float const stepPosY = piecePosY + pieceDef.sizeY * 0.5f;

            svg.elem("use")
               .attr("xlink:href", format("#{}", stepDef.id))
               .attr("transform", format("translate({} {})", stepPosX, stepPosY))
               .term();
        }
    }
};


int main(int, char*[])
{
    Vect2 constexpr pieceCellsA[] = { {0, 0}, {1, 0}, {0, 1}, {1, 1}, };
    Vect2 constexpr pieceCellsB[] = { {0, 0}, {0, 1}, };
    Vect2 constexpr pieceCellsC[] = { {0, 0}, {1, 0}, };
    Vect2 constexpr pieceCellsD[] = { {0, 0}, };

    KlotskiGrid startingGrid = {};
    startingGrid.pieces = {
        { {'A', 1}, {1, 0}, pieceCellsA, },
        { {'B', 1}, {0, 0}, pieceCellsB, },
        { {'B', 2}, {3, 0}, pieceCellsB, },
        { {'B', 3}, {0, 2}, pieceCellsB, },
        { {'B', 4}, {3, 2}, pieceCellsB, },
        { {'C', 1}, {1, 2}, pieceCellsC, },
        { {'D', 1}, {0, 4}, pieceCellsD, },
        { {'D', 2}, {1, 3}, pieceCellsD, },
        { {'D', 3}, {2, 3}, pieceCellsD, },
        { {'D', 4}, {3, 4}, pieceCellsD, },
    };

    std::cout << "initial grid:" << startingGrid << "\n";

    try {
        KlotskiSolution solution = solvePuzzle(
            startingGrid,
            [](KlotskiGrid const& grid) {
                for (auto const& piece : grid.pieces)
                    if (piece.tag == PieceTag{'A', 1})
                        return piece.position == Vect2{1, 3};
                return false;
            },
            KlotskiGrid::HorizontalSymmetry);

        std::cout << "solved grid:" << solution.grid << "\n";
        std::cout << "list of moves (" << solution.path.size() << "):\n";
        for (auto const& move : solution.path) {
            auto const& piece = startingGrid.pieces[move.pieceIndex];
            std::cout << piece.name() << move.step.toString() << " ";
        }
        std::cout << "\n";

        auto const filename = "klotski_solution.svg";
        std::ofstream svgFile(filename, std::ios::out | std::ios::binary);
        if (!svgFile.is_open()) {
            std::cerr << "could not open svg file in write mode\n";
            return 1;
        }
        KlotskiSVGRenderer{}.renderGrids(svgFile, startingGrid, solution.path);
        svgFile.close();
        return 0;
    }
    catch (std::exception const& e) {
        std::cerr << "ERROR: " << e.what() << "\n";
        return 1;
    }
}

File: python/klotski_solver.py
# SPDX-License-Identifier: MIT
# Copyright © 2023  Bilal Djelassi

from puzzle_types import Vect2, Step, PieceTag, Piece, Move, Grid, SearchTree
from svg_renderer import SVGRenderer, create_xml_element, serialize_xml_document

class SearchEdge:
    __slots__ = 'parent_index', 'move'

    def __init__(self, parent_index, move):
        self.parent_index = parent_index
        self.move = move


def solve_puzzle(initial_grid, success_condition, symmetry):
    validated = initial_grid.validate()
    if validated is None:
        raise ValueError("initial grid is invalid")
    if success_condition(initial_grid):
        return (initial_grid, [])

    search_tree = SearchTree()
    search_tree.append(initial_grid, None, validated.key(symmetry))

    while True:
        search_tree.increment_depth()
        a, b = search_tree.current_depth()

        if a == b:
            raise RuntimeError("reached end of tree, no more solutions to explore")

        for parent_index in range(a, b):
            for piece_index in range(len(initial_grid.pieces)):
                for step in Step.ALL:
                    move = Move(piece_index, step)
                    grid = search_tree.node_at(parent_index)
                    grid = grid.copy()
                    grid.apply(move)

                    validated = grid.validate()
                    if validated is None:
                        continue

                    key = validated.key(symmetry)
                    edge = SearchEdge(parent_index, move)
                    if not search_tree.append(grid, edge, key):
                        continue

                    if success_condition(grid):
                        edge = search_tree.edge_at(search_tree.last_index())
                        path = []
                        while edge is not None:
                            path.append(edge.move)
                            edge = search_tree.edge_at(edge.parent_index)
                        return (grid, list(reversed(path)))

def klotski_is_solved(grid):
    for piece in grid.pieces:
        if piece.tag == PieceTag('A', 1):
            return piece.position == Vect2(1, 3)
    return False

def format_move(grid, move):
    piece = grid.pieces[move.piece_index]
    return f"{piece.tag}{move.step.symbol}"


class CustomSVGRenderer(SVGRenderer):
    def __init__(self, grid):
        super().__init__(grid)
        self.properties = SVGRenderer.colorful_theme()

    def custom_render_grid(self, svg_element, grid, title):
        grid_size_x, grid_size_y = self.get_grid_size(False)
        svg_element.append(create_xml_element("rect", {
            "x": 0,
            "y": 0,
            "width": grid_size_x,
            "height": grid_size_y,
            "rx": "{:g}".format(self.border_radius),
            "fill": self.properties["fillColorGrid"],
            "stroke": self.properties["strokeColorGrid"],
            "stroke-width": self.properties["strokeWidthGrid"],
        }))
        if title:
            svg_element.append(create_xml_element("text", {
                "x": grid_size_x * 0.5,
                "y": self.title_height * -0.5,
                "text-anchor": "middle",
                "fill": self.properties["textColor"],
                "font-family": self.properties["fontFamily"],
                "font-size": self.properties["fontSize"],
            }, content=title))

    def custom_render_piece(self, svg_element, piece, step):
        if piece.tag.symbol == 'A':
            piece_size_x = 2 * self.unit_size + self.cell_gap
            piece_size_y = 2 * self.unit_size + self.cell_gap
        elif piece.tag.symbol == 'B':
            piece_size_x = self.unit_size
            piece_size_y = 2 * self.unit_size + self.cell_gap
        elif piece.tag.symbol == 'C':
            piece_size_x = 2 * self.unit_size + self.cell_gap
            piece_size_y = self.unit_size
        elif piece.tag.symbol == 'D':
            piece_size_x = self.unit_size
            piece_size_y = self.unit_size
        else:
            raise ValueError("cannot draw unknown piece")

        highlight = "Hi" if step else "Lo"
        piece_fill_color = self.properties[
            "fillColor{}Piece{}".format(highlight, piece.tag.symbol)]
        piece_stroke_color = self.properties[
            "strokeColor{}Piece{}".format(highlight, piece.tag.symbol)]

        piece_pos_x, piece_pos_y = self.piece_position_to_xy_coord(
            piece.position.x, piece.position.y)

        svg_element.append(create_xml_element("rect", {
            "x": piece_pos_x,
            "y": piece_pos_y,
            "width": piece_size_x,
            "height": piece_size_y,
            "rx": "{:g}".format(self.border_radius),
            "fill": piece_fill_color,
            "stroke": piece_stroke_color,
            "stroke-width": self.properties["strokeWidthPiece"],
        }))
        if step is not None:
            if step == Step.UP:
                arrow_path = "M 0 -6 L -6 4 L 6 4 Z"
            elif step == Step.DOWN:
                arrow_path = "M 0 6 L -6 -4 L 6 -4 Z"
            elif step == Step.LEFT:
                arrow_path = "M -6 0 L 4 -6 L 4 6 Z"
            elif step == Step.RIGHT:
                arrow_path = "M 6 0 L -4 -6 L -4 6 Z"
            else:
                raise ValueError("cannot draw unknown step")

            arrow_pos_x = piece_pos_x + piece_size_x * 0.5
            arrow_pos_y = piece_pos_y + piece_size_y * 0.5
            svg_element.append(create_xml_element("path", {
                "d": arrow_path,
                "fill": self.properties["arrowColor"],
                "transform": "translate({:g} {:g})".format(arrow_pos_x, arrow_pos_y),
            }))


def main():
    piece_cells_A = [Vect2.coerce(cell)
        for cell in ((0, 0), (1, 0), (0, 1), (1, 1))]
    piece_cells_B = [Vect2.coerce(cell)
        for cell in ((0, 0), (0, 1))]
    piece_cells_C = [Vect2.coerce(cell)
        for cell in ((0, 0), (1, 0))]
    piece_cells_D = [Vect2.coerce(cell)
        for cell in ((0, 0),)]

    initial_grid = Grid(4, 5, [
        Piece(('A', 1), (1, 0), piece_cells_A),
        Piece(('B', 1), (0, 0), piece_cells_B),
        Piece(('B', 2), (3, 0), piece_cells_B),
        Piece(('B', 3), (0, 2), piece_cells_B),
        Piece(('B', 4), (3, 2), piece_cells_B),
        Piece(('C', 1), (1, 2), piece_cells_C),
        Piece(('D', 1), (0, 4), piece_cells_D),
        Piece(('D', 2), (1, 3), piece_cells_D),
        Piece(('D', 3), (2, 3), piece_cells_D),
        Piece(('D', 4), (3, 4), piece_cells_D),
    ])

    print(f"initial grid:\n{initial_grid}")

    solved_grid, path = solve_puzzle(
            initial_grid,
            klotski_is_solved,
            Grid.KeySymmetry.HORIZOTAL_SYMMETRY)

    print(f"solved grid:\n{solved_grid}")

    formatted_moves = " ".join(format_move(initial_grid, move) for move in path)
    print(f"list of moves ({len(path)}):\n{formatted_moves}")

    svg = CustomSVGRenderer(initial_grid).render_grids(initial_grid, path)
    with open("klotski_solution.svg", "w") as svg_file:
        svg_file.write(serialize_xml_document(svg))

if __name__ == '__main__':
    main()
